common/src/main/kotlin/packet/Packet.kt

```diff
@@ -1,11 +1,28 @@
 package packet
 
+import io.ktor.utils.io.*
+import packet.Packet.Companion.readPacket
+
 abstract class Packet(val id: PacketId) {
-//	abstract suspend fun deserialize(readChannel: ByteReadChannel)
-//	suspend fun serialize(writeChannel: ByteWriteChannel) {
-//		writeChannel.writeByte(id.value)
-//		writePacketContent(writeChannel)
-//	}
-//
-//	protected abstract suspend fun writePacketContent(writeChannel: ByteWriteChannel)
-}
\ No newline at end of file
+	protected abstract suspend fun writePacketContent(writer: ByteWriteChannel)
+
+	companion object {
+		suspend fun ByteWriteChannel.writePacket(packet: Packet) {
+			this.writeByte(packet.id.value)
+			packet.writePacketContent(this)
+		}
+
+		suspend fun ByteReadChannel.readPacket(): Packet {
+			val packetId = PacketId(this.readByte())
+			val packetHandler = map[packetId] ?: error("unknown packet ID: ${packetId.value}")
+			return packetHandler(this)
+		}
+
+		private val map = mapOf(
+			PacketId.NODE_CREATE to ByteReadChannel::readPacketNodeCreate,
+			PacketId.GOTO to ByteReadChannel::readPacketGoTo,
+			PacketId.NAMECHANGE to ByteReadChannel::readPacketNameChange,
+			PacketId.NODE_REVEAL to ByteReadChannel::readPacketNodeReveal
+		)
```

Comment:
variables should be at the top and functions below. also "map" is not a descriptive variable name